File: pipeline.py
"""
End-to-end DaliLite pipeline orchestration.

Wires together:
  Path A: WOLF -> DP(zcut=2.0) -> DALICON(lfitz=T) -> DP(zcut=2.0)
  Path B: PARSI -> FILTER95(zcut1=1.0) -> pipe96(top 2) -> DALICON(lfitz=T) -> DP(zcut=2.0)

Both paths run in forward (query=cd1) and reverse (query=cd2) directions.
"""

import sys
import tempfile
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from wolf.wolf import wolf_compare, setup_protein, load_protein, SpatialHashGrid, NEIBORCUTOFF
from dp.dp import run_dp, DCCPResult, _format_ranges
from dalicon.dalicon import run_dalicon
from parsi.parsi import dowork_parsi, init_parsi
from filter95.filter95 import run_filter95

PIPENBEST = 2  # max entries per (cd1,cd2) key from FILTER95->DALICON


# ---------------------------------------------------------------------------
# Format converters
# ---------------------------------------------------------------------------

def dccp_to_dalicon_lines(dccp_results):
    """Convert DP DCCPResult list to DALICON input format string.

    Replaces dccp2dalicon.pl. Groups results by cd1 with END delimiters.
    """
    lines = []
    old_cd1 = None

    for r in dccp_results:
        lines.append('')  # blank line before each entry
        if r.cd1 != old_cd1:
            old_cd1 = r.cd1
            lines.append('END')
            lines.append(r.cd1)
        lines.append(f'{r.cd2}*')
        lines.append(str(r.nblock))
        # cd1 ranges in Fortran 8(i4,2x,i4) format
        lines.extend(_format_ranges(r.l1, r.r1, r.nblock))
        # cd2 ranges
        lines.extend(_format_ranges(r.l2, r.r2, r.nblock))

    lines.append('')
    lines.append('END')
    return '\n'.join(lines) + '\n'


def filter95_to_dalicon_lines(filter95_lines, zcut=1.0, nbest=PIPENBEST):
    """Convert FILTER95 output lines to DALICON input format string.

    Replaces: sort -nr | uniq | pipe96-free.pl zcut nbest
    """
    # Parse lines
    entries = []
    seen_lines = set()
    for line in filter95_lines:
        stripped = line.strip()
        if not stripped:
            continue
        # Dedup (uniq on sorted input = remove exact duplicate lines)
        if stripped in seen_lines:
            continue
        seen_lines.add(stripped)

        # Remove -99 placeholders (matching Perl s/\-99//g)
        cleaned = stripped.replace('-99', '')
        parts = cleaned.split()
        if len(parts) < 5:
            continue
        try:
            zscore = float(parts[0])
        except ValueError:
            continue

        key = parts[1]  # cd1cd2 (10 chars)
        # parts[2]=idom, parts[3]=score
        nseg = int(parts[4])
        range_vals = [int(x) for x in parts[5:5 + 4 * nseg]]
        entries.append((zscore, key, nseg, range_vals))

    # Sort by z-score descending (matching sort -nr)
    entries.sort(key=lambda x: -x[0])

    # Group by key, keep top nbest per key
    key_counts = {}
    output_lines = []
    old_cd1 = '?????'

    for zscore, key, nseg, range_vals in entries:
        key_counts[key] = key_counts.get(key, 0) + 1
        if key_counts[key] > nbest:
            continue
        if zscore < zcut:
            continue

        cd1 = key[:5].strip()
        cd2 = key[5:].strip()

        if cd1 != old_cd1:
            if old_cd1 != '?????':
                output_lines.append('END')
            output_lines.append(cd1)
            old_cd1 = cd1

        n = nseg
        output_lines.append(f'{cd2}*')
        output_lines.append(str(n))
        output_lines.append(' '.join(str(v) for v in range_vals[:2 * n]))
        output_lines.append(' '.join(str(v) for v in range_vals[2 * n:4 * n]))

    output_lines.append('END')
    output_lines.append('END')
    return '\n'.join(output_lines) + '\n'


# ---------------------------------------------------------------------------
# Pipeline paths
# ---------------------------------------------------------------------------

def run_wolf_path(query, targets, dat_dir):
    """Wolf path: WOLF -> DP(zcut=2.0) -> DALICON(lfitz=T) -> DP(zcut=2.0).

    Args:
        query: structure code to build grid for
        targets: list of structure codes to compare against
        dat_dir: path to .dat files

    Returns:
        list of DCCPResult from the final DP
    """
    dat_dir = Path(dat_dir)

    # Step 1: WOLF
    cd1_data = setup_protein(dat_dir / f'{query}.dat')
    prot1 = cd1_data[0]

    wolf_results = []
    if prot1.nseg > 2:
        grid = SpatialHashGrid()
        load_protein(grid, prot1.nseg, cd1_data[1], cd1_data[2],
                     cd1_data[3], NEIBORCUTOFF)
        for target in targets:
            result = wolf_compare(query, target, str(dat_dir),
                                  grid=grid, cd1_data=cd1_data)
            if result is not None:
                wolf_results.append(result)

    if not wolf_results:
        return []

    with tempfile.TemporaryDirectory() as tmpdir:
        tmpdir = Path(tmpdir)

        # Step 2: DP on WOLF output
        wolfitz_path = tmpdir / 'wolfitz.txt'
        with open(wolfitz_path, 'w') as f:
            for r in wolf_results:
                f.write(r.to_wolfitz_line() + '\n')

        dp_results = run_dp(str(wolfitz_path), str(dat_dir), zcut=2.0)
        if not dp_results:
            return []

        # Step 3: DCCP -> DALICON input
        dalicon_text = dccp_to_dalicon_lines(dp_results)
        dalicon_path = tmpdir / 'dalicon_input.txt'
        with open(dalicon_path, 'w') as f:
            f.write(dalicon_text)

        # Step 4: DALICON (lfitz=True by default)
        dalicon_output = run_dalicon(str(dalicon_path), str(dat_dir))
        if not dalicon_output:
            return []

        # Step 5: Final DP
        wolfitz2_path = tmpdir / 'wolfitz2.txt'
        with open(wolfitz2_path, 'w') as f:
            for line in dalicon_output:
                f.write(line + '\n')

        final_results = run_dp(str(wolfitz2_path), str(dat_dir), zcut=2.0)

    return final_results


def run_parsi_path(query, targets, dat_dir):
    """Parsi path: PARSI -> FILTER95 -> pipe96 -> DALICON -> DP.

    Args:
        query: structure code (search tree built for this)
        targets: list of structure codes to compare against
        dat_dir: path to .dat files

    Returns:
        list of DCCPResult from the final DP
    """
    dat_dir = Path(dat_dir)
    dat_dir_slash = str(dat_dir) + '/'  # filter95 needs trailing slash

    # Step 1: PARSI
    init_parsi()
    all_refine_lines = []
    cd1_cache = {}
    for target in targets:
        lines = dowork_parsi(query, target, str(dat_dir),
                             lfirstonly=True, cd1_cache=cd1_cache)
        all_refine_lines.extend(lines)

    if not all_refine_lines:
        return []

    # Step 2: FILTER95
    filter95_output = run_filter95(all_refine_lines, dat_dir_slash, zcut1=1.0)
    if not filter95_output:
        return []

    # Step 3: pipe96 -> DALICON input
    dalicon_text = filter95_to_dalicon_lines(filter95_output, zcut=1.0,
                                             nbest=PIPENBEST)

    with tempfile.TemporaryDirectory() as tmpdir:
        tmpdir = Path(tmpdir)

        # Step 4: DALICON
        dalicon_path = tmpdir / 'dalicon_input.txt'
        with open(dalicon_path, 'w') as f:
            f.write(dalicon_text)

        dalicon_output = run_dalicon(str(dalicon_path), str(dat_dir))
        if not dalicon_output:
            return []

        # Step 5: Final DP
        wolfitz_path = tmpdir / 'wolfitz.txt'
        with open(wolfitz_path, 'w') as f:
            for line in dalicon_output:
                f.write(line + '\n')

        final_results = run_dp(str(wolfitz_path), str(dat_dir), zcut=2.0)

    return final_results


# ---------------------------------------------------------------------------
# Top-level entry point
# ---------------------------------------------------------------------------

def compare_pair(cd1, cd2, dat_dir):
    """Full comparison of two structures.

    Runs wolf and parsi paths in both forward (query=cd1) and reverse
    (query=cd2) directions, accumulates all DCCP results.

    Args:
        cd1: first structure code
        cd2: second structure code
        dat_dir: path to .dat files

    Returns:
        list of DCCPResult (self-comparisons filtered out)
    """
    all_results = []
    targets = [cd1, cd2]

    # Forward direction: query=cd1
    all_results.extend(run_wolf_path(cd1, targets, dat_dir))
    all_results.extend(run_parsi_path(cd1, targets, dat_dir))

    # Reverse direction: query=cd2
    all_results.extend(run_wolf_path(cd2, targets, dat_dir))
    all_results.extend(run_parsi_path(cd2, targets, dat_dir))

    # Filter out self-comparisons
    all_results = [r for r in all_results if r.cd1 != r.cd2]

    return all_results

File: dali_rust/dali-core/src/pipeline.rs
//! Pipeline orchestration module.
//!
//! Wires together:
//!   Path A (Wolf): WOLF → DP(zcut=2.0) → DALICON(lfitz=T) → DP(zcut=2.0)
//!   Path B (Parsi): PARSI → FILTER95(zcut=1.0) → pipe96(top 2) → DALICON(lfitz=T) → DP(zcut=2.0)
//!
//! Both paths run in forward (query=cd1) and reverse (query=cd2) directions.
//! Self-comparisons are filtered from final output.

use std::collections::{HashMap, HashSet};

use rayon::prelude::*;

use crate::types::{AlignmentBlock, DccpEntry, SearchHit, mask_protein};
use crate::store::ProteinStore;
use crate::numerics::compute_transform;
use crate::wolf;
use crate::dp;
use crate::dalicon::{self, DaliconRecord};
use crate::parsi;
use crate::filter95::{self, Filter95Entry};

/// Max entries per (cd1,cd2) key from FILTER95 → DALICON.
const PIPENBEST: usize = 2;

// ---------------------------------------------------------------------------
// Format converters
// ---------------------------------------------------------------------------

/// Convert DccpEntry blocks to DaliconRecord format.
///
/// Flattens AlignmentBlocks into the values array that `convert_prealignment`
/// expects: values[0..2*nblock] = cd1 (l1,r1) pairs, values[2*nblock..4*nblock] = cd2 (l2,r2) pairs.
fn dccp_to_dalicon_record(entry: &DccpEntry) -> DaliconRecord {
    let nblock = entry.blocks.len();
    let mut values = Vec::with_capacity(nblock * 4);
    // cd1 ranges
    for b in &entry.blocks {
        values.push(b.l1 as i32);
        values.push(b.r1 as i32);
    }
    // cd2 ranges
    for b in &entry.blocks {
        values.push(b.l2 as i32);
        values.push(b.r2 as i32);
    }
    DaliconRecord {
        cd1: entry.cd1.clone(),
        cd2: entry.cd2.clone(),
        nblock,
        values,
    }
}

/// Convert Filter95Entry list to DaliconRecords via pipe96 filtering.
///
/// Matches: `sort -nr | uniq | pipe96-free.pl zcut nbest`
/// - Sort by zscore descending
/// - Keep top `nbest` per (cd1+cd2) key
/// - Filter by z-score cutoff
fn filter95_to_dalicon_records(
    entries: &[Filter95Entry],
    zcut: f64,
    nbest: usize,
) -> Vec<DaliconRecord> {
    let mut sorted: Vec<&Filter95Entry> = entries.iter().collect();
    sorted.sort_by(|a, b| b.zscore.partial_cmp(&a.zscore).unwrap_or(std::cmp::Ordering::Equal));

    let mut key_counts: HashMap<String, usize> = HashMap::new();
    let mut records = Vec::new();

    for e in sorted {
        if e.zscore < zcut {
            continue;
        }
        let key = format!("{}{}", e.cd1, e.cd2);
        let count = key_counts.entry(key).or_insert(0);
        *count += 1;
        if *count > nbest {
            continue;
        }

        records.push(DaliconRecord {
            cd1: e.cd1.clone(),
            cd2: e.cd2.clone(),
            nblock: e.nseg,
            values: e.ranges.clone(),
        });
    }

    records
}

// ---------------------------------------------------------------------------
// Pipeline paths
// ---------------------------------------------------------------------------

/// Wolf path: WOLF → DP(zcut=2.0) → DALICON(lfitz=T) → DP(zcut=2.0).
pub fn run_wolf_path(
    query: &str,
    targets: &[&str],
    store: &ProteinStore,
) -> Vec<DccpEntry> {
    let dat_dir = store.dat_dir();

    // Step 1: WOLF — setup query protein and build spatial hash grid
    let cd1_path = format!("{}/{}.dat", dat_dir, query);
    let cd1_data = match wolf::setup_protein(&cd1_path) {
        Ok(d) => d,
        Err(_) => return Vec::new(),
    };

    if cd1_data.protein.nseg <= 2 {
        return Vec::new();
    }

    let mut grid = wolf::spatial_hash::SpatialHashGrid::new();
    wolf::load_protein(&mut grid, &cd1_data);

    let mut wolf_results = Vec::new();
    for &target in targets {
        let cd2_path = format!("{}/{}.dat", dat_dir, target);
        let cd2_data = match wolf::setup_protein(&cd2_path) {
            Ok(d) => d,
            Err(_) => continue,
        };
        if let Some(result) = wolf::wolf_compare(&cd1_data, &cd2_data, &grid) {
            wolf_results.push(result);
        }
    }

    if wolf_results.is_empty() {
        return Vec::new();
    }

    // Step 2: DP on WOLF output
    let alignments: Vec<(String, String, Vec<AlignmentBlock>)> = wolf_results
        .into_iter()
        .map(|r| (r.cd1, r.cd2, r.blocks))
        .collect();

    let dp_results = dp::run_dp(&alignments, store, 2.0);
    if dp_results.is_empty() {
        return Vec::new();
    }

    // Step 3: DALICON on DP output (lfitz=true)
    let records: Vec<DaliconRecord> = dp_results.iter().map(dccp_to_dalicon_record).collect();
    let dalicon_output = dalicon::run_dalicon(&records, store);

    if dalicon_output.is_empty() {
        return Vec::new();
    }

    // Step 4: Final DP
    let final_alignments: Vec<(String, String, Vec<AlignmentBlock>)> = dalicon_output
        .into_iter()
        .map(|(cd1, cd2, _nblock, blocks)| (cd1, cd2, blocks))
        .collect();

    dp::run_dp(&final_alignments, store, 2.0)
}

/// Parsi path: PARSI → FILTER95(zcut=1.0) → pipe96(top 2) → DALICON(lfitz=T) → DP(zcut=2.0).
pub fn run_parsi_path(
    query: &str,
    targets: &[&str],
    store: &ProteinStore,
) -> Vec<DccpEntry> {
    let dat_dir = store.dat_dir();

    // Step 1: PARSI
    let mut all_hits = Vec::new();
    let mut cd1_cache: Option<parsi::ParsiCd1Cache> = None;

    for &target in targets {
        let hits = parsi::dowork_parsi(
            query, target, dat_dir, None, true, &mut cd1_cache,
        );
        all_hits.extend(hits);
    }

    if all_hits.is_empty() {
        return Vec::new();
    }

    // Step 2: FILTER95
    let filter_results = filter95::run_filter95(&all_hits, store, Some(1.0));
    if filter_results.is_empty() {
        return Vec::new();
    }

    // Step 3: pipe96 → DALICON input records
    let records = filter95_to_dalicon_records(&filter_results, 1.0, PIPENBEST);
    if records.is_empty() {
        return Vec::new();
    }

    // Step 4: DALICON (lfitz=true)
    let dalicon_output = dalicon::run_dalicon(&records, store);
    if dalicon_output.is_empty() {
        return Vec::new();
    }

    // Step 5: Final DP
    let final_alignments: Vec<(String, String, Vec<AlignmentBlock>)> = dalicon_output
        .into_iter()
        .map(|(cd1, cd2, _nblock, blocks)| (cd1, cd2, blocks))
        .collect();

    dp::run_dp(&final_alignments, store, 2.0)
}

// ---------------------------------------------------------------------------
// Top-level entry point
// ---------------------------------------------------------------------------

/// Full comparison of two structures (serial).
///
/// Runs wolf and parsi paths in both forward (query=cd1) and reverse
/// (query=cd2) directions. Self-comparisons are filtered out.
pub fn compare_pair_serial(
    cd1: &str,
    cd2: &str,
    store: &ProteinStore,
) -> Vec<DccpEntry> {
    let targets: Vec<&str> = vec![cd1, cd2];
    let mut all_results = Vec::new();

    // Forward direction: query=cd1
    all_results.extend(run_wolf_path(cd1, &targets, store));
    all_results.extend(run_parsi_path(cd1, &targets, store));

    // Reverse direction: query=cd2
    all_results.extend(run_wolf_path(cd2, &targets, store));
    all_results.extend(run_parsi_path(cd2, &targets, store));

    // Filter out self-comparisons
    all_results.retain(|r| r.cd1 != r.cd2);

    all_results
}

/// Full comparison of two structures (parallel via Rayon).
///
/// Runs 4 independent paths concurrently:
///   1. Forward wolf  (cd1 → [cd1,cd2])
///   2. Forward parsi (cd1 → [cd1,cd2])
///   3. Reverse wolf  (cd2 → [cd1,cd2])
///   4. Reverse parsi (cd2 → [cd1,cd2])
///
/// Each path has its own mutable state (DaliconCd1State, ParsiCd1Cache).
/// Only the ProteinStore is shared (thread-safe via RwLock + Arc).
pub fn compare_pair(
    cd1: &str,
    cd2: &str,
    store: &ProteinStore,
) -> Vec<DccpEntry> {
    let targets: Vec<&str> = vec![cd1, cd2];

    // Define the 4 independent tasks as (query, path_type) pairs
    let tasks: Vec<(&str, bool)> = vec![
        (cd1, true),   // forward wolf
        (cd1, false),  // forward parsi
        (cd2, true),   // reverse wolf
        (cd2, false),  // reverse parsi
    ];

    let mut all_results: Vec<DccpEntry> = tasks
        .into_par_iter()
        .flat_map(|(query, is_wolf)| {
            if is_wolf {
                run_wolf_path(query, &targets, store)
            } else {
                run_parsi_path(query, &targets, store)
            }
        })
        .collect();

    // Filter out self-comparisons
    all_results.retain(|r| r.cd1 != r.cd2);

    all_results
}

// ---------------------------------------------------------------------------
// Database search (one-to-many)
// ---------------------------------------------------------------------------

/// One-to-many structural search with query-side amortization.
///
/// Runs WOLF and/or PARSI paths (forward only: query→targets).
/// Filters self-comparisons, applies z-score cutoff, deduplicates by
/// keeping the highest z-score per (cd1, cd2) pair, and sorts descending.
///
/// When `skip_wolf` is true, only the PARSI path runs — useful when an
/// upstream prefilter (e.g. FoldSeek) has already narrowed candidates.
pub fn search_database(
    query: &str,
    targets: &[&str],
    store: &ProteinStore,
    z_cut: f64,
    skip_wolf: bool,
) -> Vec<DccpEntry> {
    let mut all_results = Vec::new();

    if !skip_wolf {
        all_results.extend(run_wolf_path(query, targets, store));
    }
    all_results.extend(run_parsi_path(query, targets, store));

    // Filter self-comparisons
    all_results.retain(|r| r.cd1 != r.cd2);

    // Filter by z-score
    all_results.retain(|r| r.zscore >= z_cut);

    // Deduplicate: keep highest z-score per (cd1, cd2) pair
    let mut best_map: HashMap<(String, String), DccpEntry> = HashMap::new();
    for entry in all_results {
        let key = (entry.cd1.clone(), entry.cd2.clone());
        let replace = match best_map.get(&key) {
            Some(existing) => entry.zscore > existing.zscore,
            None => true,
        };
        if replace {
            best_map.insert(key, entry);
        }
    }

    let mut results: Vec<DccpEntry> = best_map.into_values().collect();
    results.sort_by(|a, b| b.zscore.partial_cmp(&a.zscore).unwrap_or(std::cmp::Ordering::Equal));
    results
}

// ---------------------------------------------------------------------------
// Iterative search (search + mask loop)
// ---------------------------------------------------------------------------

/// Iterative multi-domain detection: search → take best → mask → repeat.
///
/// Each round runs `search_database`, takes the top hit, computes the
/// structural transform, masks aligned residues from the query, and repeats
/// until no hit exceeds thresholds or the query becomes too small.
pub fn iterative_search(
    query: &str,
    targets: &[&str],
    store: &ProteinStore,
    min_aligned: usize,
    min_zscore: f64,
    gap_tolerance: usize,
    max_rounds: usize,
    skip_wolf: bool,
) -> Vec<SearchHit> {
    let mut hits = Vec::new();
    let mut current_query = query.to_string();

    for round in 0..max_rounds {
        let target_refs: Vec<&str> = targets.iter().copied().collect();
        let results = search_database(&current_query, &target_refs, store, min_zscore, skip_wolf);

        if results.is_empty() {
            break;
        }

        let best = &results[0];

        // Count aligned residues
        let n_aligned: usize = best.blocks.iter()
            .map(|b| (b.r1 - b.l1 + 1) as usize)
            .sum();
        if n_aligned < min_aligned {
            break;
        }

        // Load proteins for transform computation
        let query_prot = match store.get_protein(&current_query) {
            Ok(p) => p,
            Err(_) => break,
        };
        let template_prot = match store.get_protein(&best.cd2) {
            Ok(p) => p,
            Err(_) => break,
        };

        // Compute rotation + translation
        let (rotation, translation) = match compute_transform(&query_prot.ca, &template_prot.ca, &best.blocks) {
            Some((u, t)) => (u, t),
            None => break,
        };

        // Expand blocks to residue-level alignment pairs (1-based)
        let mut alignments: Vec<(usize, usize)> = Vec::new();
        let mut aligned_indices: HashSet<usize> = HashSet::new(); // 0-based query indices

        for b in &best.blocks {
            let len = (b.r1 - b.l1 + 1) as usize;
            for k in 0..len {
                let q_idx = (b.l1 as usize - 1) + k; // 0-based
                let t_idx = (b.l2 as usize - 1) + k;
                if q_idx < query_prot.nres && t_idx < template_prot.nres {
                    alignments.push((q_idx + 1, t_idx + 1)); // 1-based
                    aligned_indices.insert(q_idx);
                }
            }
        }

        hits.push(SearchHit {
            cd2: best.cd2.clone(),
            zscore: best.zscore,
            score: best.score,
            rmsd: best.rmsd,
            nblock: best.blocks.len(),
            blocks: best.blocks.clone(),
            rotation,
            translation,
            alignments,
            round,
        });

        // Compute mask set: aligned indices + gap-bridged neighbors
        let effective_gap = gap_tolerance.max(n_aligned * 5 / 100);
        let mut mask_set = aligned_indices.clone();

        // Bridge small gaps between aligned regions
        if !aligned_indices.is_empty() {
            let mut sorted_aligned: Vec<usize> = aligned_indices.iter().copied().collect();
            sorted_aligned.sort();
            for window in sorted_aligned.windows(2) {
                let gap = window[1] - window[0];
                if gap > 1 && gap <= effective_gap + 1 {
                    for idx in (window[0] + 1)..window[1] {
                        mask_set.insert(idx);
                    }
                }
            }
        }

        // Compute keep_indices: all residues NOT in mask
        let keep_indices: Vec<usize> = (0..query_prot.nres)
            .filter(|i| !mask_set.contains(i))
            .collect();

        if keep_indices.len() < min_aligned {
            break;
        }

        // Mask and store
        let new_code = format!("{}_r{}", query, round);
        let masked = mask_protein(&query_prot, &keep_indices, &new_code);

        if masked.nseg < 3 {
            // Too few SSEs for WOLF/PARSI to work meaningfully
            break;
        }

        if store.add_protein(masked).is_err() {
            break;
        }
        current_query = new_code;
    }

    hits
}
